google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentConfigurationPanel.java

```diff
@@ -31,6 +44,99 @@
+
+  public AppEngineDeploymentConfigurationPanel() {
+    versionIdField
+        .getEmptyText()
+        .setText(GctBundle.message("appengine.flex.version.placeholder.text"));
+
+    promoteCheckbox.setSelected(PROMOTE_DEFAULT);
+    stopPreviousVersionCheckbox.setSelected(STOP_PREVIOUS_VERSION_DEFAULT);
+    stopPreviousVersionCheckbox.setEnabled(STOP_PREVIOUS_VERSION_DEFAULT);
+
+    promoteInfoLabel.setHyperlinkText(
+        GctBundle.getString("appengine.promote.info.label.beforeLink") + " ",
+        GctBundle.getString("appengine.promote.info.label.link"),
+        "");
+    promoteInfoLabel.addHyperlinkListener(new BrowserOpeningHyperLinkListener());
+    promoteInfoLabel.setHyperlinkTarget(GctBundle.getString("appengine.promoteinfo.url"));
+
+    promoteCheckbox.addItemListener(
+        event -> {
+          boolean isPromoteSelected = ((JCheckBox) event.getItem()).isSelected();
+
+          stopPreviousVersionCheckbox.setEnabled(isPromoteSelected);
+          stopPreviousVersionCheckbox.setSelected(isPromoteSelected);
+        });
+
+    appEngineCostWarningLabel.setHyperlinkText(
+        GctBundle.getString("appengine.flex.deployment.cost.warning.beforeLink"),
+        GctBundle.getString("appengine.flex.deployment.cost.warning.link"),
+        " " + GctBundle.getString("appengine.flex.deployment.cost.warning.afterLink"));
+    appEngineCostWarningLabel.addHyperlinkListener(new BrowserOpeningHyperLinkListener());
+    appEngineCostWarningLabel.setHyperlinkTarget(CloudSdkAppEngineHelper.APP_ENGINE_BILLING_URL);
+
+    projectSelector.addProjectSelectionListener(applicationInfoPanel::refresh);
+
+    projectSelector.addModelListener(
+        new TreeModelAdapter() {
+          @Override
+          public void treeStructureChanged(TreeModelEvent event) {
+            // projects have finished loading
+            refreshApplicationInfoPanel();
+          }
+        });
+  }
+
+  public void resetEditorFrom(@NotNull AppEngineDeploymentConfiguration configuration) {
+    promoteCheckbox.setSelected(configuration.isPromote());
+    versionIdField.setText(configuration.getVersion());
+    stopPreviousVersionCheckbox.setSelected(configuration.isStopPreviousVersion());
+    projectSelector.setText(configuration.getCloudProjectName());
+
+    refreshApplicationInfoPanel();
+  }
+
+  public void applyEditorTo(@NotNull AppEngineDeploymentConfiguration configuration) {
+    configuration.setVersion(versionIdField.getText());
+    configuration.setPromote(promoteCheckbox.isSelected());
+    configuration.setCloudProjectName(projectSelector.getText());
+    configuration.setStopPreviousVersion(stopPreviousVersionCheckbox.isSelected());
+  }
+
+  /**
+   * Sets the project / version to allow the deployment line items to be decorated with additional
+   * identifying data. See {@link AppEngineRuntimeInstance#getDeploymentName}.
+   */
+  public void setDeploymentProjectAndVersion(DeploymentSource deploymentSource) {
```

Comment:
leaving this out unless you / @patflynn feel strongly about it. imo this can get noisy